daml-foundations/daml-ghc/src/DA/Daml/GHC/Compiler/Convert.hs

```diff
@@ -566,7 +566,11 @@ convertExpr env0 e = do
     go env (VarIs "getFieldPrim") (Type (isStrLitTy -> Just name) : Type record : Type _field : args) = fmap (, args) $ do
         record' <- convertType env record
         pure $ ETmLam (varV1, record') $ ERecProj (fromTCon record') (mkField $ unpackFS name) $ EVar varV1
-    go env (VarIs "getField") (Type (isStrLitTy -> Just name) : Type recordType : Type _fieldType : _dict : record : args) = fmap (, args) $ do
+    -- NOTE(MH): We only inline `getField` for record types. This is required
+    -- for contract keys. Projections on sum-of-records types have to through
+    -- the type class for `getField`.
+    go env (VarIs "getField") (Type (isStrLitTy -> Just name) : Type recordType@(TypeCon recordTyCon _) : Type _fieldType : _dict : record : args)
+        | Just [_] <- tyConDataCons_maybe recordTyCon = fmap (, args) $ do
```

Comment:
Didn't we have a helper function to test if something was a single constructor?

Reply:
The helper function works on the data constructor level, not the type constructor level.